File: python/Sum of Left Leaves.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def array_to_tree(arr):
    if not arr:
        return None

    root = TreeNode(arr[0])
    queue = [root]
    i = 1

    while queue and i < len(arr):
        node = queue.pop(0)
        if arr[i] != 'null':
            node.left = TreeNode(arr[i])
            queue.append(node.left)
        i += 1
        if i < len(arr) and arr[i] != 'null':
            node.right = TreeNode(arr[i])
            queue.append(node.right)
        i += 1

    return root

def sumOfLeftLeaves(root):
    def dfs(node, is_left):
        if not node:
            return 0
        if not node.left and not node.right:
            return node.val if is_left else 0
        return dfs(node.left, True) + dfs(node.right, False)

    return dfs(root, False)

# Example usage
input_array = [3, 9, 20, 'null', 'null', 15, 7]
root = array_to_tree(input_array)
print(sumOfLeftLeaves(root))
'''
A leaf is a node with no children. A left leaf is a leaf that is the left child of another node.

 

Example 1:


Input: root = [3,9,20,null,null,15,7]
Output: 24
Explanation: There are two left leaves in the binary tree, with values 9 and 15 respectively.
Example 2:

Input: root = [1]
Output: 0'''

File: JAVA/java/src/SumofLeftLeaves.java
import java.util.LinkedList;
import java.util.Queue;

class TreeNode4 {
    int val;
    TreeNode4 left, right;

    TreeNode4(int val) {
        this.val = val;
    }
}

public class SumofLeftLeaves {
    public static TreeNode4 arrayToTree(Object[] arr) {
        if (arr == null || arr.length == 0 || arr[0] == null)
            return null;

        Queue<TreeNode4> queue = new LinkedList<>();
        TreeNode4 root = new TreeNode4((int) arr[0]);
        queue.offer(root);
        int i = 1;

        while (!queue.isEmpty() && i < arr.length) {
            TreeNode4 currNode = queue.poll();

            if (arr[i] != null) {
                currNode.left = new TreeNode4((int) arr[i]);
                queue.offer(currNode.left);
            }
            i++;

            if (i < arr.length && arr[i] != null) {
                currNode.right = new TreeNode4((int) arr[i]);
                queue.offer(currNode.right);
            }
            i++;
        }
        return root;
    }

    public static int sumOfLeftLeaves(TreeNode4 root) {
        return dfs(root, false);
    }

    private static int dfs(TreeNode4 node, boolean isLeft) {
        if (node == null) return 0;
        if (node.left == null && node.right == null) return isLeft ? node.val : 0;
        int leftSum = dfs(node.left, true);
        int rightSum = dfs(node.right, false);
        return leftSum + rightSum;
    }

    public static void main(String[] args) {
        Object[] inputArray = {3, 9, 20, null, null, 15, 7};
        TreeNode4 root = arrayToTree(inputArray);
        System.out.println(sumOfLeftLeaves(root));
    }
}

